src/libstore/local-store.cc

```diff
@@ -1081,17 +1084,34 @@ Path LocalStore::addToStore(const string & name, const Path & _srcPath,
     bool recursive, HashType hashAlgo, PathFilter & filter, RepairFlag repair)
 {
     Path srcPath(absPath(_srcPath));
+    Hash h = recursive ? hashPath(hashAlgo, srcPath, filter).first : hashFile(hashAlgo, srcPath);
+    Path dstPath = makeFixedOutputPath(recursive, h, name);
+    addTempRoot(dstPath);
+    if (repair || !isValidPath(dstPath)) {
+
+        PathLocks outputLock({dstPath});
 
-    /* Read the whole path into memory. This is not a very scalable
-       method for very large paths, but `copyPath' is mainly used for
-       small files. */
-    StringSink sink;
-    if (recursive)
-        dumpPath(srcPath, sink, filter);
-    else
-        sink.s = make_ref<std::string>(readFile(srcPath));
+        if (repair || !isValidPath(dstPath)) {
+
+            if (pathExists(dstPath)) deletePath(dstPath);
+
+            copyPath(srcPath, dstPath, filter, recursive);
+
+            canonicalisePathMetaData(dstPath, -1);
+            HashResult hash = hashPath(htSHA256, dstPath);
```

Comment:
So we copy then hash? How does this differ from the `hashPath` call above?

Reply:
I suspect this `hashPath` is redundant. Assuming that `canonicalisePathMetaData` doesn't affect the hash, it seems like we should be able to use the hash `h` computed above.

Reply:
@Mathnerd314, do you recall why this second hash was needed?

Reply:
See the comment [over here](https://github.com/NixOS/nix/pull/1754/files#diff-dcbcac55a634031f9cc73707da6e4b18L1052). The first hash computes the output path, the second is the SHA256 of the NAR serialization which is used for validity checking. I guess the code could re-use the hash in the case of recursive SHA256, but hashing isn't that expensive.